ocaml/xapi/xapi_network.ml

```diff
@@ -329,3 +329,62 @@ let with_networks_attached_for_vm ~__context ?host ~vm f =
         error "Caught %s while detaching networks" (string_of_exn e)
     end;
     raise e
+
+let assert_can_add_purpose ~__context ~network ~current porpoise =
+  let sop (*string-of-porpoise*) = Record_util.network_purpose_to_string in
+  let reject str =
+    raise (Api_errors.Server_error (Api_errors.network_incompatible_purposes, [ (sop porpoise); str ]))
+  in
+  if (List.mem `himn current) then (
+    info "Existing network purpose %s is not compatible with any other purpose." (sop `himn);
+    reject "any"
+  );
+  if (List.mem `unmanaged current) then (
+    info "Existing network purpose %s is not compatible with any other purpose." (sop `unmanaged);
+    reject "any"
+  );
+
+  let assert_no_net_has_bad_porpoise bads =
+    (* Sadly we can't use Db.Network.get_refs_where because the expression
+     * type doesn't allow searching for a value inside a list. *)
+    Db.Network.get_all ~__context |>
+    List.iter (fun nwk ->
+      Db.Network.get_purposes ~__context ~self:nwk |>
+      List.iter (fun suspect -> if (List.mem suspect bads) then
+        info "Cannot set new network purpose %s when there is a network with purpose %s" (sop porpoise) (sop suspect);
+        reject (sop suspect)
+      )
+    )
+  in
+  match porpoise with
+  | `himn | `unmanaged -> if current <> [] then (
+      info "New network purpose %s is not compatible with any other purpose" (sop porpoise);
+      reject "any"
+    )
+  | `management -> if (List.mem `guest current) then (
+      info "Combining network purposes %s and %s is inadvisable for security reasons, but not prevented." (sop `management) (sop `guest);
+      reject (sop `guest)
+    )
+  | `guest -> if (List.mem `management current) then (
+      info "Combining network purposes %s and %s is inadvisable for security reasons, but not prevented." (sop `management) (sop `guest);
+      reject (sop `management)
+    )
+  | `storage -> ()
+  | `nbd -> assert_no_net_has_bad_porpoise [`insecure_nbd]
+  | `insecure_nbd -> assert_no_net_has_bad_porpoise [`nbd]
+
+let add_purpose ~__context ~network ~purpose =
+  let self = network in
+  let current = Db.Network.get_purposes ~__context ~self in
+  if not (List.mem purpose current) then (
+    assert_can_add_purpose ~__context ~network ~current purpose;
+    Db.Network.set_purposes ~__context ~self ~value:(purpose::current)
+  )
+
+let remove_purpose ~__context ~network ~purpose =
+  let self = network in
+  let current = Db.Network.get_purposes ~__context ~self in
+  if (List.mem purpose current) then (
+    let porpoises = List.filter (fun porpoise -> porpoise<>purpose) current in
+    Db.Network.set_purposes ~__context ~self ~value:(porpoises)
+  )
```

Comment:
These functions are fine, just interested: did you decide against using `Db.Network.add/remove_purpose` because that's not idempotent? Just asking because I'm not sure whether they throw an exception for already existing/removed keys, and that would be good to know.